internal/server/polling/server.go

```diff
@@ -105,66 +113,161 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 	s.log.Info("start polling", "namespace", resource.Namespace, "name", resource.Name)
 
 	if secret == nil {
-		return fmt.Errorf("secret is not defined")
+		err := fmt.Errorf("secret is not defined")
+		s.log.Error(err, "failed to get secret")
+		return err
 	}
 
+	s.log.Info("fetching terraform object", "namespace", resource.Namespace, "name", resource.Name)
 	tf, err := s.getTerraformObject(ctx, resource)
 	if err != nil {
-		return fmt.Errorf("failed to get Terraform object %s/%s: %w", resource.Namespace, resource.Name, err)
+		s.log.Error(err, "failed to get terraform object", "namespace", resource.Namespace, "name", resource.Name)
+		return fmt.Errorf("failed to get terraform object %s/%s: %w", resource.Namespace, resource.Name, err)
 	}
 
+	s.log.Info("fetching source object", "namespace", resource.Namespace, "name", resource.Name)
 	source, err := s.getSource(ctx, tf)
 	if err != nil {
-		return fmt.Errorf("failed to get Source object: %w", err)
+		s.log.Error(err, "failed to get source object")
+		return fmt.Errorf("failed to get source object: %w", err)
 	}
 
+	s.log.Info("initializing git provider", "url", source.Spec.URL)
 	gitProvider, repo, err := provider.FromURL(
 		source.Spec.URL,
 		provider.WithLogger(s.log),
 		provider.WithToken("api-token", string(secret.Data["token"])),
 	)
 
 	if err != nil {
+		s.log.Error(err, "failed to get git provider")
 		return fmt.Errorf("failed to get git provider: %w", err)
 	}
 
+	s.log.Info("listing pull requests")
 	prs, err := gitProvider.ListPullRequests(ctx, repo)
 	if err != nil {
+		s.log.Error(err, "failed to list pull requests")
 		return fmt.Errorf("failed to list pull requests: %w", err)
 	}
 
-	return s.reconcile(ctx, tf, source, prs)
+	s.log.Info("reconciling pull requests")
+	return s.reconcile(ctx, tf, source, prs, gitProvider)
 }
 
-func (s *Server) reconcile(ctx context.Context, original *infrav1.Terraform, source *sourcev1.GitRepository, prs []provider.PullRequest) error {
-	// List Terraform objects, created by the branch planner.
-	tfList, err := s.listTerraformObjects(ctx, original.Namespace, map[string]string{
-		bpconfig.LabelKey: bpconfig.LabelValue,
-	})
-	if err != nil {
-		return fmt.Errorf("failed to list Terraform objects: %w", err)
-	}
+func (s *Server) reconcile(ctx context.Context, original *infrav1.Terraform, source *sourcev1.GitRepository, prs []provider.PullRequest, gitProvider provider.Provider) error {
+	s.log.Info("starting reconciliation ...")
```

Comment:
Is it worth to create a new local logger here and use that? That would add the minimal context to all log messages.
```go
log := s.log.WithValues("terraform", fmt.Sprintf("%s/%s", original.Namespace, original.Name))

log.Info("starting reconciliation ...")

// ...

log.Info("mapping PR", "PR number", pr.Number)

// ...
if err != nil {
  log.Error(err, "failed to reconcile Terraform object")
}
```

Or would it be a mess to use this logic?

Reply:
It's fine to do so for this one and other functions, later, to be consistent.